Chore/remove table calculation field type

joaoviana opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

Closes: <!-- reference the related issue e.g. #150 -->

### Description:
<!-- Add a description of the changes proposed in the pull request. -->

<!-- Even better add a screenshot / gif / loom -->

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

joaoviana commented on packages/backend/src/queryCompiler.ts:
```diff
@@ -1,17 +1,14 @@
 import {
     AdditionalMetric,
-    CompiledCustomDimension,
```
unused imports

joaoviana commented on packages/backend/src/services/ProjectService/ProjectService.ts:
```diff
@@ -3052,7 +3054,7 @@ export class ProjectService extends BaseService {
+        const allFilterableFields: FilterableDimension[] = [];
         const filterIndexMap: Record<string, number> = {};
 
         allFilters.forEach((filterSet) => {
```
`allFilters` are `FilterableDimension`s

joaoviana commented on packages/common/src/types/field.ts:
```diff
@@ -178,17 +178,16 @@ export enum TableCalculationType {
     format?: CustomFormat;
     type?: TableCalculationType;
+    // TODO: Check if this needs to be required - it's not in the metricQuery compiled
+    fieldType?: FieldType.TABLE_CALCULATION;
```
❓ I believe I should not add this here - it was only in `TableCalculationField`

ZeRego replied:
correct, we don't need this 👍 
We can also remove `fieldType` from `Field` as it was added alongside `TableCalculationField`

joaoviana replied:
Done here: https://github.com/lightdash/lightdash/pull/10062/commits/2f8ca5ed0574d660d829e607fe9a0b6c3eaeb9ce

joaoviana commented on packages/common/src/types/field.ts:
```diff
@@ -479,6 +446,33 @@ export interface Metric extends Field {
     requiredAttributes?: Record<string, string | string[]>; // Required attributes for the dimension this metric is based on
 }
 
+export const isFilterableDimension = (
```
Moved this here so we could use `isMetric` in `isFilterableField`

joaoviana commented on packages/common/src/utils/filters.ts:
```diff
@@ -98,10 +98,6 @@ export const getFilterGroupItemsPropertyName = (
     return 'and';
 };
 
-export const filterableDimensionsOnly = (
```
Not used anywhere

joaoviana commented on packages/common/src/utils/filters.ts:
```diff
@@ -139,6 +135,8 @@ export const getFilterTypeFromItem = (item: FilterableItem): FilterType => {
         default: {
             return assertUnreachable(
+                // @ts-expect-error - TODO: Fix this
+                // TODO: Fix this
```
❓ Need help with this - this appeared after changes to `FilterableItem` type

ZeRego replied:
Do we need to handle custom dimensions? I was hoping to do this in the other ticket.

```
const type = isCustomBinDimension ? DimensionType.STRING : isCustomSqlDimension ? item.dimensionType : item.type;
switch (type)
...
```

joaoviana replied:
Nope, we dont... `FilterableField` does not contain `CustomDimension`s...

joaoviana replied:
https://github.com/lightdash/lightdash/pull/10062/commits/ec23c3cb8dd1acde51d62ea9d3f4ecf62a6e4a36

This removes the reference of `itemsMap` in `FilterableField` because that was include `Field`, and field as a `type: string` (**not an enum**) - hence why that `assertUnreachable` was complaining about a `string`

joaoviana commented on packages/frontend/src/components/Explorer/CustomMetricModal/FilterForm.tsx:
```diff
@@ -38,36 +39,40 @@ export const FilterForm: FC<{
 
-    const dimensions = Object.values(fieldsMap).filter(isDimension);
+    const dimensionsMap = getFilterableDimensionsFromItemsMap(itemsMap);
+    const dimensions = Object.values(dimensionsMap);
```
Custom metric modal filters were only using dimensions

joaoviana commented on packages/frontend/src/components/Explorer/CustomMetricModal/index.tsx:
```diff
@@ -323,7 +323,7 @@ export const CustomMetricModal = () => {
                                 <FiltersProvider
                                     projectUuid={projectUuid}
-                                    fieldsMap={fieldsMap}
+                                    itemsMap={fieldsMap}
```
perhaps the filtering in: https://github.com/lightdash/lightdash/pull/10062/files#r1599738715 should be done before here? 
```
itemsMap={dimensionsMap}
```
?

ZeRego replied:
ideally yes

joaoviana replied:
done here: https://github.com/lightdash/lightdash/pull/10062/commits/6ec4497e9149073f79ff1042af5a0b6aa6e14d46

joaoviana commented on packages/frontend/src/components/Explorer/FiltersCard/useFieldsWithSuggestions.ts:
```diff
@@ -54,9 +60,10 @@ export const useFieldsWithSuggestions = ({
                 >((acc, cal) => {
-                    const tableCalculationFilters: TableCalculationField = {
+                    // TODO: Verify if this is OK
+                    const tableCalculationFilters = {
```
❓ I removed the explicit typing. what do you think?

ZeRego replied:
Can we remove the loop as we don't need to convert?

joaoviana replied:
done: https://github.com/lightdash/lightdash/pull/10062/commits/0797a01809c674faf5c70decac835d9a8e0252e8
thanks for the suggestion 🙏

joaoviana commented on packages/frontend/src/components/common/Filters/FilterInputs/index.tsx:
```diff
@@ -246,8 +246,8 @@ export const getConditionalRuleLabel = (
     filterRule: ConditionalRule,
-    field: FilterableField,
-    filterableFields: FilterableField[],
+    field: FilterableDimension,
```
Only used in Dashboard filters

joaoviana commented on packages/frontend/src/components/common/Filters/index.tsx:
```diff
@@ -41,12 +43,28 @@ type Props = {
     isEditMode: boolean;
 };
 
+const getInvalidFilterRules = (
```
Moved this util to this component because it's only used here

joaoviana commented:
For me to review 
- Filters in Filter form
- Filters in Dashboards (specially availableFilters endpoint - play around with tile targets)
- Filter dashboard to...
- Creating filter rules for any
- Propagation of filter rules when exploring from here...

joaoviana commented:
After testing: 
- Only issue I've found was with adding Table Calculations as filters on an explore. Testing here: https://lightdash-pr-10062.onrender.com/projects/3675b69e-8324-4110-bdca-059031aa8da3/saved/6ec076a7-314b-42cf-aa2b-fedf6b5818fa/view

ZeRego commented on packages/backend/src/queryBuilder.mock.ts:
```diff
@@ -1036,7 +1036,7 @@ export const METRIC_QUERY_WITH_TABLE_CALCULATION_FILTER: CompiledMetricQuery = {
                     },
                     operator: FilterOperator.EQUALS,
-                    values: [],
+                    values: ['my value'],
```
This unit test was wrong from the beginning. 
The table calculation id would not match with the filter target id (`table_calculation_{table name}`) causing the test to fall into this condition.
`if (!field) return '1=1';`

joaoviana commented:
Tested the following again: https://github.com/lightdash/lightdash/pull/10062#issuecomment-2111896629 and it's working as expected
- Added TC
- Added it as a filter
- Then changed its type
- Added again as a filter

Tested Dashboard filters and the `availableFilters` endpoint
Confirmed as well that I don't see table calculations' filters where they shouldn't be - for example, in Dashboards & Custom metric modal

joaoviana commented on packages/backend/src/database/migrations/20240515102618_remove_table_calculation_filter_prefix.ts:
```diff
@@ -0,0 +1,26 @@
+        .select('saved_queries_version_uuid')
+        .whereRaw("jsonb_path_exists(filters, '$.tableCalculations')");
+
+    await knex('saved_queries_versions')
```
```suggestion
 if (versionsWithTableCalculationFilters.length > 0) {
    await knex('saved_queries_versions')
```

I think this should be added, what do you think? ^

ZeRego replied:
yeah, we can add it :)

magnew commented on packages/common/src/types/field.ts:
```diff
@@ -199,26 +195,16 @@ export type TableCalculationMetadata = {
-    sql: string;
 }
 
 // This type check is a little fragile because it's based on
```
I think this comment is no longer relevant.
